src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php

```diff
@@ -21,6 +21,8 @@
 
 /**
  * @author Jordi Boggiano <j.boggiano@seld.be>
+ *
+ * @template TObject
```

Comment:
Shouldn't this be moved to the method PHPdoc?

Reply:
By putting the `@template`, don't we ask implementors to declare which types they will implement, `DenormalizerInterface<foo>`?
Putting the annotation just binds the value of the $type argument to the return type, which looks enough to me, and more accurate actually.
I'm no expert so I might be wrong.

Reply:
> By putting the @template on the class, don't we ask implementors to declare which types they will implement, DenormalizerInterface<foo>?

Yes, which makes sense, as you know what you will return when implementing the interface (or `mixed` if you don't).

> Putting the annotation on the method just binds the value of the $type argument to the return type, which looks enough to me, and more accurate actually.
Purely from a typing perspective I don't think that makes sense, as it would imply that you could have multiple methods returning different types within a single class. So whenever you called the generic method you'd also have to specify which method you're expecting to call.

> I'm no expert so I might be wrong.
Same.

Reply:
Well, you get the type depending on what is passed in the format argument. It is not the class-level that decide of the generic type (which would make it impossible to properly type the caller btw). It is the caller (passing the format) that decides that.

As is, this PR hurts typing instead of improving it.

Reply:
> It is not the class-level that decide of the generic type

Why not? Maybe I do things differently, but when I create a denormalizer for a type `MyType`, it makes sense to me to specify that this is a denormalizer for the type `MyType`. The annotation then tells the static analysis that if you pass the class-string `MyType::class` as the argument, you're getting a denormalizer that returns `MyType`. The compatible types are determined by the class implementing the interface.

Reply:
That's not the class-level annotation that does that (try defining that annotation of more generic normalizers that support several types and decide based on the `$format` argument, and you'll see why your proposal is broken).

Reply:
I see what you mean, thanks for explaining. Do you see any way to fix this, or is the denormalizer doomed to be untyped?

Reply:
Moving `@template` to the method fixes the issue.

Reply:
@wouterj No it doesn't, because then psalm will output the following:

> psalm: InvalidReturnType: The declared return type '(TObject:fn-symfony\component\serializer\normalizer\denormalizerinterface::denormalize as object)&App\Serializer\Foo' for App\Serializer\FooNormalizer::denormalize is incorrect, got 'App\Serializer\Foo

I guess the serializer is impossible to properly type as it stands, so I'm closing this issue.